### PR TITLE
feat(export-to-language) add Rust support as a target language for export RUST-1155

### PR DESCRIPTION
**Pull request by abr-egn**
## Description
This adds a Rust target for the language transpilation, enabling export for find and aggregate pipelines.
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context
Rust is a relatively new language rapidly gaining traction in industry usage, and has an officially supported mongodb driver; this provides an easy path to working with that driver for Compass users.
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Open Questions
Rust does not support some constructs referenced in the export machinery (DBRefs, Decimal128), I couldn't tell if there was any better way to flag that than not implementing the relevant templates.

## Dependents
None.

## Types of changes
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by addaleax** [comment]
Thank you @abr-egn! :slightly_smiling_face: I’ve opened COMPASS-5527 to track reviewing this on our side, we’ll hopefully get to that soon, and kicked off CI here :+1:

**Comment by lerouxb** [comment]
Please just have a look here: https://github.com/mongodb-js/compass/blob/main/packages/compass-export-to-language/src/components/editor/editor.jsx#L5-L11

That got missed recently with the ruby fix. Throws some runtime error if you select a language that's not loaded there.

**Comment by abr-egn** [comment]
I've updated the editor imports - wouldn't have found that, thanks!

I'm afraid I couldn't make much out of the test failures; they didn't seem at all related to this PR?

**Comment by addaleax** [comment]
> I'm afraid I couldn't make much out of the test failures; they didn't seem at all related to this PR?

Yeah, that might happen sometimes. If test failures are looking unrelated, you can ignore them and see if they stick out to us, since we’ll probably take a look before merging anyway :)

File: packages/bson-transpilers/test/yaml/bson-methods.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/bson-methods.yaml:39]: I think we need a `.to_string()` here after `"test code"`

File: packages/bson-transpilers/test/yaml/bson-methods.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/bson-methods.yaml:54]: We could use `Bson::JavaScriptCode` here, though I'm not sure if that matches the expected type/usage.

File: packages/bson-transpilers/test/yaml/driver-syntax.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/driver-syntax.yaml:412]: we'll probably want to use `Document` as a generic parameter to `collection` here since I don't think the compiler can infer the type otherwise in this example.

File: packages/bson-transpilers/test/yaml/native.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/native.yaml:938]: I think we'll need to use the `bson::DateTime::now()` here, since this example requires `chrono`, which we're trying to move away from.

File: packages/bson-transpilers/test/yaml/native.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/native.yaml:1051]: Unfortunately we don't currently have a way to do this with `bson::DateTime`, though we probably should. I filed https://jira.mongodb.org/browse/RUST-1198 to cover the work for that, maybe in the meantime we just omit these examples?

File: packages/bson-transpilers/test/yaml/native.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/native.yaml:1109]: need `.to_string()` for both of these. ditto for the other regex examples

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/partial.yaml:95]: can we use `Bson::JavaScriptCode` here?

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/partial.yaml:191]: ditto using `bson::DateTime` instead of a chrono type

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (patrickfreed) [inline packages/bson-transpilers/test/yaml/partial.yaml:221]: I think we want `to_hex` here

File: packages/bson-transpilers/test/yaml/bson-methods.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/bson-methods.yaml]: Done, and similarly for uses of `Bson::JavaScriptCode()`.

File: packages/bson-transpilers/test/yaml/bson-methods.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/bson-methods.yaml:54]: This is testing the ability to take a code construction (i.e. `Bson::JavaScriptCode("test code".to_string())`) and extract the code string by applying an attribute/accessor, but because we rely on destructuring here there isn't one to apply.

File: packages/bson-transpilers/test/yaml/driver-syntax.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/driver-syntax.yaml]: Good catch, done.

File: packages/bson-transpilers/test/yaml/native.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/native.yaml]: Updated.

File: packages/bson-transpilers/test/yaml/native.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/native.yaml]: Actually, we can pull the same trick as js - emit the date as a string and parse from that.  More direct constructors would still certainly be a good to have.

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/partial.yaml:95]: This is testing "partial" constructions, i.e. transpilations of fragmentary input.  In this case, it's testing translating the code "symbol" without any arguments.  Because for Rust we have to switch between `JavaScriptCode` and `JavaScriptCodeWithScope` based on whether the scope argument is present, the symbol itself has to be empty.  I don't think this matters much because it's just testing nonsense.

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/partial.yaml]: Done.

File: packages/bson-transpilers/test/yaml/partial.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/partial.yaml]: Done.

**Comment by mcasimir** [comment]
@abr-egn @patrickfreed can you ping us when this PR will be ready for review by DevTools? Thanks!

File: packages/bson-transpilers/test/yaml/driver-syntax.yaml
Comment (abr-egn) [reply packages/bson-transpilers/test/yaml/driver-syntax.yaml]: Done.

File: packages/bson-transpilers/symbols/rust/templates.yaml
Comment (abr-egn) [reply packages/bson-transpilers/symbols/rust/templates.yaml:478]: This name is how it's used; I agree that `CodeScopeScopeTemplate` would have made more sense with the naming conventions.  There doesn't seem to be any support for a `CodeScopeCodeTemplate`.

File: packages/bson-transpilers/symbols/rust/templates.yaml
Comment (abr-egn) [reply packages/bson-transpilers/symbols/rust/templates.yaml]: I ... honestly have no idea what past-me was thinking here.  Updated to your suggestion.

File: packages/bson-transpilers/symbols/rust/templates.yaml
Comment (abr-egn) [reply packages/bson-transpilers/symbols/rust/templates.yaml]: Whoops, missed that one.  Fixed.

File: packages/bson-transpilers/symbols/rust/templates.yaml
Comment (abr-egn) [reply packages/bson-transpilers/symbols/rust/templates.yaml:212]: Looks much nicer, done.

File: packages/bson-transpilers/symbols/rust/templates.yaml
Comment (abr-egn) [reply packages/bson-transpilers/symbols/rust/templates.yaml]: I don't think it's base64, but there's inconsistency in how the other languages handle it - javascript treats it as a buffer object and converts *that* to base64, csharp treats it as an ASCII string, and java treats it as a UTF-8 string.  Given all that I think you're right that it's best to hold off for now; removed what was there.

**Review by patrickfreed — APPROVED** [review]
LGTM from the Rust end!

**Review by alenakhineika — APPROVED** [review]

**Pull request merged**
